### PR TITLE
CORE-915 Dead man's switch example

### PR DESCRIPTION
**Pull request by erer1243**
wip of dead man's switch example. I'm sure there are better ways to implement some parts of it :^). Figuring there is a better way to refer to a single member of a participant class, and a better way to 'die' / opt out of a contract than just sleeping.

### CODE REVIEW

File: examples/dead-mans-switch-nft/index.rsh [hunk 1]
```diff
@@ -0,0 +1,74 @@
+'reach 0.1';
+'use strict';
+
+export const main = Reach.App(() => {
+  const Creator = Participant('Creator', {
+    transferTo: Fun([], Address),
+  });
+  const Owner = ParticipantClass('Owner', {
```
Comment (jeapostrophe) [inline examples/dead-mans-switch-nft/index.rsh:8]: Don't use participant classes, ever :)

File: examples/dead-mans-switch-nft/index.rsh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+  // const ownerOnly = (currentOwner, fn) =>
+  //   Owner.only(() => { if (this == currentOwner) { fn(interact); } });
+
+  var owner = Creator;
+  invariant(true);
+  while (true) {
+    vNFT.owner.set(owner);
+
+    if (owner == Creator) {
```
Comment (jeapostrophe) [inline examples/dead-mans-switch-nft/index.rsh:28]: This doesn't mean what you think it means... the way you have it the creator can never be a "normal" owner. You should not have any `if` like this in the program

File: examples/dead-mans-switch-nft/index.rsh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  const vNFT = View('NFT', { owner: Address });
+
+  init();
+
+  Creator.publish();
+
+  // const ownerOnly = (currentOwner, fn) =>
+  //   Owner.only(() => { if (this == currentOwner) { fn(interact); } });
+
+  var owner = Creator;
```
Comment (jeapostrophe) [inline examples/dead-mans-switch-nft/index.rsh:23]: First, I think you should use `parallelReduce`

Second, the idea is that there's always an "owner" and an "heir". If the owner doesn't ping, then the heir can claim it and become the new owner (setting up the next heir).

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: examples/dead-mans-switch-nft/index.rsh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    .api(O.ping,
+      () => check(this == owner),
+      () => 0,
+      (k) => {
+        check(this == owner);
+        k(null);
+        return [owner, heir];
+      }
+    )
+    .timeout(relativeTime(3), () => {
```
Comment (jeapostrophe) [inline examples/dead-mans-switch-nft/index.rsh:37]: That's too little... it should be a parameter of the Creator

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: examples/dead-mans-switch-nft/index.rsh (hunk 4 above)
Comment (erer1243) [reply examples/dead-mans-switch-nft/index.rsh:37]: Got it

File: examples/dead-mans-switch-nft/index.rsh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      const [[nextHeir], k] =
+        call(O.setNextHeir)
+        .assume((nextHeir) => check(this != nextHeir));
```
[the remark this answers is not in the document]
Comment (erer1243) [reply examples/dead-mans-switch-nft/index.rsh:40]: Whoops, thanks.

File: examples/dead-mans-switch-nft/index.rsh [hunk 6]
```diff
@@ -34,11 +36,11 @@ export const main = Reach.App(() => {
         return [owner, heir];
```
Comment (jeapostrophe) [inline examples/dead-mans-switch-nft/index.rsh:36]: You could add `owner != heir` into the invariant

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

**Review by jeapostrophe — APPROVED** [review]

**Pull request merged**